File: clients/typescript/src/model/vehicle.ts
import { Quest } from "./quest";
import { Vec2Double } from "./vec2-double";
import { Stream } from "../stream";

/**
 * A vehicle
 */
export class Vehicle {
    /**
     * Current position (center)
     */
    position: Vec2Double
    /**
     * Velocity vector
     */
    velocity: Vec2Double
    /**
     * Speed of wheels
     */
    speed: number
    /**
     * Rotation speed (radians/second)
     */
    rotationSpeed: number
    /**
     * Current rotation
     */
    rotation: number
    /**
     * Vehicle type index
     */
    typeIndex: number
    /**
     * Current quest, if any
     */
    quest: Quest | null
    /**
     * Remaining fuel
     */
    fuel: number

    constructor(position: Vec2Double, velocity: Vec2Double, speed: number, rotationSpeed: number, rotation: number, typeIndex: number, quest: Quest | null, fuel: number) {
        this.position = position;
        this.velocity = velocity;
        this.speed = speed;
        this.rotationSpeed = rotationSpeed;
        this.rotation = rotation;
        this.typeIndex = typeIndex;
        this.quest = quest;
        this.fuel = fuel;
    }

    /**
     * Read Vehicle from input stream
     */
    static async readFrom(stream: Stream): Promise<Vehicle> {
        let position;
        position = await Vec2Double.readFrom(stream);
        let velocity;
        velocity = await Vec2Double.readFrom(stream);
        let speed;
        speed = await stream.readDouble();
        let rotationSpeed;
        rotationSpeed = await stream.readDouble();
        let rotation;
        rotation = await stream.readDouble();
        let typeIndex;
        typeIndex = await stream.readInt();
        let quest;
        if (await stream.readBool()) {
            quest = await Quest.readFrom(stream);
        } else {
            quest = null;
        }
        let fuel;
        fuel = await stream.readDouble();
        return new Vehicle(position, velocity, speed, rotationSpeed, rotation, typeIndex, quest, fuel)
    }

    /**
     * Write Vehicle to output stream
     */
    async writeTo(stream: Stream) {
        let position = this.position;
        await position.writeTo(stream);
        let velocity = this.velocity;
        await velocity.writeTo(stream);
        let speed = this.speed;
        await stream.writeDouble(speed);
        let rotationSpeed = this.rotationSpeed;
        await stream.writeDouble(rotationSpeed);
        let rotation = this.rotation;
        await stream.writeDouble(rotation);
        let typeIndex = this.typeIndex;
        await stream.writeInt(typeIndex);
        let quest = this.quest;
        if (quest === null) {
            await stream.writeBool(false);
        } else {
            await stream.writeBool(true);
            await quest.writeTo(stream);
        }
        let fuel = this.fuel;
        await stream.writeDouble(fuel);
    }
}

File: clients/ruby/model/vehicle.rb
require './model/quest'
require './model/vec2_double'

module Model

# A vehicle
class Vehicle
    # Current position (center)
    attr_accessor :position
    # Velocity vector
    attr_accessor :velocity
    # Speed of wheels
    attr_accessor :speed
    # Rotation speed (radians/second)
    attr_accessor :rotation_speed
    # Current rotation
    attr_accessor :rotation
    # Vehicle type index
    attr_accessor :type_index
    # Current quest, if any
    attr_accessor :quest
    # Remaining fuel
    attr_accessor :fuel

    def initialize(position, velocity, speed, rotation_speed, rotation, type_index, quest, fuel)
        @position = position
        @velocity = velocity
        @speed = speed
        @rotation_speed = rotation_speed
        @rotation = rotation
        @type_index = type_index
        @quest = quest
        @fuel = fuel
    end

    # Read Vehicle from input stream
    def self.read_from(stream)
        position = Model::Vec2Double.read_from(stream)
        velocity = Model::Vec2Double.read_from(stream)
        speed = stream.read_double()
        rotation_speed = stream.read_double()
        rotation = stream.read_double()
        type_index = stream.read_int()
        if stream.read_bool()
            quest = Model::Quest.read_from(stream)
        else
            quest = nil
        end
        fuel = stream.read_double()
        Vehicle.new(position, velocity, speed, rotation_speed, rotation, type_index, quest, fuel)
    end

    # Write Vehicle to output stream
    def write_to(stream)
        @position.write_to(stream)
        @velocity.write_to(stream)
        stream.write_double(@speed)
        stream.write_double(@rotation_speed)
        stream.write_double(@rotation)
        stream.write_int(@type_index)
        if @quest.nil?
            stream.write_bool(false)
        else
            stream.write_bool(true)
            @quest.write_to(stream)
        end
        stream.write_double(@fuel)
    end

    def to_s
        string_result = "Vehicle { "
        string_result += "position: "
        string_result += @position.to_s
        string_result += ", "
        string_result += "velocity: "
        string_result += @velocity.to_s
        string_result += ", "
        string_result += "speed: "
        string_result += @speed.to_s
        string_result += ", "
        string_result += "rotation_speed: "
        string_result += @rotation_speed.to_s
        string_result += ", "
        string_result += "rotation: "
        string_result += @rotation.to_s
        string_result += ", "
        string_result += "type_index: "
        string_result += @type_index.to_s
        string_result += ", "
        string_result += "quest: "
        if @quest.nil?
            string_result += "nil"
        else
            string_result += @quest.to_s
        end
        string_result += ", "
        string_result += "fuel: "
        string_result += @fuel.to_s
        string_result += " }"
        string_result
    end

    def to_str
        to_s
    end
end

end